tests/unit/httpd_test.cc

```diff
@@ -268,28 +286,35 @@ SEASTAR_TEST_CASE(test_json_path) {
     path1.set(*route, [res1] (const_req req) {
         (*res1) = true;
         BOOST_REQUIRE_EQUAL(req.param["param1"], "value1");
+        BOOST_REQUIRE_EQUAL(req.get_path_param("param1"), "value1");
         return "";
     });
 
     path2.set(*route, [res2] (const_req req) {
         (*res2) = true;
-        BOOST_REQUIRE_EQUAL(req.param["param1"], "value2");
-        BOOST_REQUIRE_EQUAL(req.param["param2"], "text1");
+        BOOST_REQUIRE_EQUAL(req.param["param1"], "value4+value4%20value4");
+        BOOST_REQUIRE_EQUAL(req.get_path_param("param1"), "value4+value4 value4");
+
+        BOOST_REQUIRE_EQUAL(req.param["param2"], "text4%2Btext4");
+        BOOST_REQUIRE_EQUAL(req.get_path_param("param2"), "text4+text4");
```

Comment:
So I don't understand why this is useful. Unlike my patch in https://github.com/scylladb/seastar/pull/2162, here you decided not to fix `param["param2"]` and instead provide a new, fixed, function get_path_param. Why? When will it **ever** make sense to use the unfixed one?

Reply:
In particular, IIRC (I'll need to confirm this), this broken param[] API is exactly what ScyllaDB used and led to the bug https://github.com/scylladb/scylladb/pull/18091 where an undecoded path was used. So in this patch you leave the broken API and only users who remember to switch to the new API will get the fix. We will see this bug 5 years from now as new users will use the old API and wonder why it's broken. So, why should we leave a broken API and not fix existing software already using the old API and unknowingly have this bug?

Reply:
I have now deprecated the old API and adopted some of the tests from your PR.

However, I did not port the path decoding inside `parse_query_param` for now. This changes the url seastar uses for routing to be the decoded url. I am curious if you think that we should fix this in place (changing the existing http routing behaviour) or if we should keep that as is. If we do path decode the return value of `parse_query_param`, some further changes are needed to pass both the raw and the decoded urls into `routes::handle` to allow us to keep the raw path parameters in `req->param`. Let me know what you think.

Also, I haven't yet made optimized for the decoded (non-buggy) use case yet. `request::get_path_param` is still suboptimal compared to `param[]` as it requires path decoding on each access. For this, I was thinking that we could either introduce a second `httpd::parameters decoded_params;` inside `http::request` or have a second `unordered_map` inside `http::parameters`. Let me know if you have a preference for either.

Reply:
Also, I left this comment on your PR which is related to why I haven't yet implemented the path decoding inside `parse_query_param`: https://github.com/scylladb/seastar/pull/2162#discussion_r1551811108

Reply:
I don't remember now if I changed that routing code because it was exercised in a test, or because I needed in the Scylla use case. I'll test your patch later with my Scylla reproducer.
I think that most Web servers when they get the URL `/some/dir/file%20name` will treat it just like `/some/dir/file name` and it should also work for "routing" - if it doesn't the first user to try to use this feature with encoded characters will be suprised :-(
I guess you're worried about backward compatibility with old code that might want the old unencoded path?

Reply:
> I guess you're worried about backward compatibility with old code that might want the old unencoded path?

Yes, I am unsure if I can have such a breaking change in a patch version change or if it needs to be gated with `SEASTAR_API_LEVEL` or similar. But you know better than me about this, so if you think we can fix this in place, I am happy to make that change as well.

But for redpanda, no, I do not think routing on decoded urls will cause any issues after looking through the API we expose.

However, if we want to keep exposing undecoded path parameters with `param[]` while starting to route based on the decoded path, the implementation will be a bit more complex as both the decoded and undecoded url will have to be passed down from `connection::generate_reply` all the way to `param_matcher::match`. But perhaps that's the least of our concerns here.